packages/flutter/lib/src/services/restoration.dart

```diff
@@ -0,0 +1,888 @@
+  /// example, the owner of a bucket that was originally claimed from the
+  /// [RestorationManager.rootBucket] must dispose that bucket and claim a new
+  /// bucket from the new [RestorationManager.rootBucket]. Once the new bucket
+  /// is claimed, owners should restore their state according to the data stored
+  /// in the new bucket.
+  ///
+  /// A call to [decommission] must always be followed by a call to [dispose].
+  void decommission() {
+    _parent?._dropChild(this);
+    _parent = null;
+    _performDecommission();
+  }
+
+  void _performDecommission() {
+    _manager = null;
+    notifyListeners();
+    _visitChildren((RestorationBucket bucket) {
+      bucket._performDecommission();
+    });
+  }
+
+  // Get and store values.
+
+  /// Returns the value that is currently stored in the bucket under the
+  /// provided `id`.
+  ///
+  /// Returns null if nothing is stored under that id.
+  ///
+  /// See also:
+  ///
+  ///  * [put], which stores a value in the bucket.
+  ///  * [remove], which removes a value from the bucket.
+  ///  * [containsValue], which checks whether any value is stored under a given
+  ///    [RestorationId].
+  P get<P>(RestorationId id) {
+    assert(id != null);
+    return _rawValues[id.value] as P;
+  }
+
+  /// Stores the provided `value` under the provided `id` in the bucket.
+  ///
+  /// Any value that has previously been stored under that id is overwritten
+  /// with the new value. The provided `value` must be serializable with the
+  /// [StandardMessageCodec].
+  ///
+  /// Null values will be stored in the bucket as-is. To remove a value, use
+  /// [remove].
+  ///
+  /// See also:
+  ///
+  ///  * [get], which retrieves a stored value from the bucket.
+  ///  * [remove], which removes a value from the bucket.
+  ///  * [containsValue], which checks whether any value is stored under a given
+  ///    [RestorationId].
+  void put<P>(RestorationId id, P value) {
+    assert(id != null);
+    assert(debugIsSerializableForRestoration(value));
+    if (_rawValues[id.value] != value || !_rawValues.containsKey(id.value)) {
+      _rawValues[id.value] = value;
+      _manager?.scheduleSerialization();
+    }
+  }
+
+  /// Deletes the value currently stored under the provided `id` from the
+  /// bucket.
+  ///
+  /// The value removed from the bucket is returned. If no value was stored
+  /// under that id, null is returned.
+  ///
+  /// See also:
+  ///
+  ///  * [get], which retrieves a stored value from the bucket.
+  ///  * [put], which stores a value in the bucket.
+  ///  * [containsValue], which checks whether any value is stored under a given
+  ///    [RestorationId].
+  P remove<P>(RestorationId id) {
+    assert(id != null);
+    final bool needsUpdate = _rawValues.containsKey(id.value);
+    final P result = _rawValues.remove(id.value) as P;
+    if (_rawValues.isEmpty) {
+      _rawData.remove(_valuesMapKey);
+    }
+    if (needsUpdate) {
+      _manager?.scheduleSerialization();
```

Comment:
Similar comment for this one - if I've arrived in a state without a manager, this won't actually get removed right?

Reply:
It will get removed from the local _rawData and hence is no longer in the bucket.

Since without a manager the _rawData wasn't persisted anywhere there's no other place to remove it from.

Reply:
What I was looking for here is what happens when:

1. You persist some data
2. You get unhooked from the manager
3. You try to update the data

And then your application restarts and you have data you thought you removed.